README.md
# Minimalist ToC

A minimalist table of contents in Javascript.

## Presentation

This script `minimalist-toc.js` allow you to print a table of contents in your page. To use it, just initialize it by declaring:

- an id corresponding to the content of the page,
- an id corresponding to the location of the table of contents,
- eventually some options.

By default, the table of contents:

- display a title (`Table of contents`) inside `h2` tag,
- use an ordered list (`ol`),
- search all heading levels except `h1` (so `h2`, `h3`, `h4`, `h5`, `h6`).

This settings can be overwritten by using the options.

If your titles do not have an id, the script generates one, otherwise it uses the existing id.

Finally, as the name suggests, it is a minimalist table of contents. Only the script generating the ToC is provided. There is no defined style. So you can define your own styles without having to overwrite existing ones.

## Example

```javascript
const Minimalist_TOC = new TableOfContent();

Minimalist_TOC.init("entry-content", "entry-toc", {
  headings: ["h2", "h3"],
  title: "My custom title",
  titleTag: "p",
  listType: "ul",
});
```

Here I redefine all the options, but you can change just one or even zero if you want.

## Acknowledgments

The `slugify` function was not written by me. So, thanks to the various contributors of [this gist](https://gist.github.com/codeguy/6684588). I only made some changes to suit my needs.

## License

This project is open source and available under the [MIT License](https://github.com/ArmandPhilippot/minimalist-toc/blob/master/LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/ArmandPhilippot/minimalist-toc/compare/v1.0.0...v1.0.1) (2021-01-05)


### Docs

* rename script and specify its name in README + add author ([39965fa](https://github.com/ArmandPhilippot/minimalist-toc/commit/39965fa5582013344b71b68af95c18ec7867ef41))

## 1.0.0 (2021-01-05)


### Changed

* initial release ([184dc09](https://github.com/ArmandPhilippot/minimalist-toc/commit/184dc09e1dfce368761129a97240c39134decfd1))


toc.js
/**
 * Convert a text into a slug or id.
 *
 * @see: https://gist.github.com/codeguy/6684588#gistcomment-3332719
 *
 * @param {String} text Text to slugify.
 */
const slugify = (text) => {
  return text
    .toString()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .trim()
    .replace(/\s+/g, "-")
    .replace(/[^\w\-]+/g, "-")
    .replace(/\-\-+/g, "-")
    .replace(/^-|-$/g, "");
};

class TableOfContent {
  constructor() {
    this.source = "";
    this.target = "";
    this.options = {
      headings: ["h2", "h3", "h4", "h5", "h6"],
      title: "Table of contents",
      titleTag: "h2",
      listType: "ol",
    };
  }

  setSource(source) {
    if (source && typeof source === "string") {
      this.source = document.getElementById(source);
    }
  }

  setTarget(target) {
    if (target && typeof target === "string") {
      this.target = document.getElementById(target);
    }
  }

  setOptions(options) {
    if (options) {
      if (options.headings && options.headings instanceof Array) {
        this.options.headings = options.headings;
      }
      if (options.title && typeof options.title === "string") {
        this.options.title = options.title;
      }
      if (options.titleTag && typeof options.titleTag === "string") {
        this.options.titleTag = options.titleTag;
      }
      if (options.listType && typeof options.listType === "string") {
        this.options.listType = options.listType;
      }
    }
  }

  getHeadingList() {
    return this.source.querySelectorAll(this.options.headings);
  }

  createTitleMarkup() {
    return (
      "<" +
      this.options.titleTag +
      ">" +
      this.options.title +
      "</" +
      this.options.titleTag +
      ">"
    );
  }

  startListNode() {
    return "<" + this.options.listType + ">";
  }

  endListNode() {
    return "</" + this.options.listType + ">";
  }

  determineLevel(currentHeading) {
    return this.options.headings.indexOf(currentHeading);
  }

  calculateLevelDiff(currentLevel, previousLevel) {
    return currentLevel - previousLevel;
  }

  addSlug(heading) {
    heading.id = heading.id ? heading.id : slugify(heading.innerText);
    return heading.id;
  }

  createTocMarkup() {
    let levelDiff;
    let headingList = this.getHeadingList();
    let i = 0;
    let markup = "";
    let currentLevel;
    let previousLevel = currentLevel;

    markup += this.createTitleMarkup();
    markup += this.startListNode();

    for (i; i < headingList.length; i++) {
      currentLevel = this.determineLevel(headingList[i].localName);
      levelDiff = this.calculateLevelDiff(currentLevel, previousLevel);

      if (levelDiff < 0) {
        while (levelDiff < 0) {
          markup += this.endListNode();
          levelDiff++;
        }
      } else if (levelDiff > 0) {
        while (levelDiff > 0) {
          markup += this.startListNode();
          levelDiff--;
        }
      }

      if (i !== 0) {
        markup += "</li>";
      }

      markup += "<li>";
      markup +=
        '<a href="#' +
        this.addSlug(headingList[i]) +
        '">' +
        headingList[i].innerText +
        "</a>";

      previousLevel = currentLevel;
    }
    markup += "</li>";
    markup += this.endListNode();

    return markup;
  }

  printToc() {
    if (this.source && this.target) {
      this.target.innerHTML = this.createTocMarkup();
    }
  }

  /**
   * Initialize the table of content.
   *
   * @param {String} source The container id to look for headings.
   * @param {String} target The container id to display the table of content.
   * @param {Object} options Options.
   *
   * Options list :
   * - headings: an array of headings to retrieve. Default : `['h2', 'h3', 'h4', 'h5', 'h6']`
   * - title: the title to print before the table of contents. Default: `Table of contents`.
   * - titleTag: the title tag. Default: `h2`.
   * - listType: the list type (`ul` or `ol`). Default: `ol`.
   */
  init(source, target, options) {
    this.setSource(source);
    this.setTarget(target);
    this.setOptions(options);
    this.printToc();
  }
}

const Minimalist_TOC = new TableOfContent();

/*
 * Here an example to init the table of contents without title:
 * - `content` is the id of the page content
 * - `toc` is the id where to place the table of contents.
 *
 * Minimalist_TOC.init('content', 'toc', {
	title: '',
});
 */


minimalist-toc.js
/**
 * Minimalist ToC
 *
 * A minimalist table of contents in Javascript.
 *
 * @author Armand Philippot <contact@armandphilippot.com>
 */

/**
 * Convert a text into a slug or id.
 *
 * @see: https://gist.github.com/codeguy/6684588#gistcomment-3332719
 *
 * @param {String} text Text to slugify.
 */
const slugify = (text) => {
  return text
    .toString()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .trim()
    .replace(/\s+/g, "-")
    .replace(/[^\w\-]+/g, "-")
    .replace(/\-\-+/g, "-")
    .replace(/^-|-$/g, "");
};

class TableOfContent {
  constructor() {
    this.source = "";
    this.target = "";
    this.options = {
      headings: ["h2", "h3", "h4", "h5", "h6"],
      title: "Table of contents",
      titleTag: "h2",
      listType: "ol",
    };
  }

  setSource(source) {
    if (source && typeof source === "string") {
      this.source = document.getElementById(source);
    }
  }

  setTarget(target) {
    if (target && typeof target === "string") {
      this.target = document.getElementById(target);
    }
  }

  setOptions(options) {
    if (options) {
      if (options.headings && options.headings instanceof Array) {
        this.options.headings = options.headings;
      }
      if (options.title && typeof options.title === "string") {
        this.options.title = options.title;
      }
      if (options.titleTag && typeof options.titleTag === "string") {
        this.options.titleTag = options.titleTag;
      }
      if (options.listType && typeof options.listType === "string") {
        this.options.listType = options.listType;
      }
    }
  }

  getHeadingList() {
    return this.source.querySelectorAll(this.options.headings);
  }

  createTitleMarkup() {
    return (
      "<" +
      this.options.titleTag +
      ">" +
      this.options.title +
      "</" +
      this.options.titleTag +
      ">"
    );
  }

  startListNode() {
    return "<" + this.options.listType + ">";
  }

  endListNode() {
    return "</" + this.options.listType + ">";
  }

  determineLevel(currentHeading) {
    return this.options.headings.indexOf(currentHeading);
  }

  calculateLevelDiff(currentLevel, previousLevel) {
    return currentLevel - previousLevel;
  }

  addSlug(heading) {
    heading.id = heading.id ? heading.id : slugify(heading.innerText);
    return heading.id;
  }

  createTocMarkup() {
    let levelDiff;
    let headingList = this.getHeadingList();
    let i = 0;
    let markup = "";
    let currentLevel;
    let previousLevel = currentLevel;

    markup += this.createTitleMarkup();
    markup += this.startListNode();

    for (i; i < headingList.length; i++) {
      currentLevel = this.determineLevel(headingList[i].localName);
      levelDiff = this.calculateLevelDiff(currentLevel, previousLevel);

      if (levelDiff < 0) {
        while (levelDiff < 0) {
          markup += this.endListNode();
          levelDiff++;
        }
      } else if (levelDiff > 0) {
        while (levelDiff > 0) {
          markup += this.startListNode();
          levelDiff--;
        }
      }

      if (i !== 0) {
        markup += "</li>";
      }

      markup += "<li>";
      markup +=
        '<a href="#' +
        this.addSlug(headingList[i]) +
        '">' +
        headingList[i].innerText +
        "</a>";

      previousLevel = currentLevel;
    }
    markup += "</li>";
    markup += this.endListNode();

    return markup;
  }

  printToc() {
    if (this.source && this.target) {
      this.target.innerHTML = this.createTocMarkup();
    }
  }

  /**
   * Initialize the table of content.
   *
   * @param {String} source The container id to look for headings.
   * @param {String} target The container id to display the table of content.
   * @param {Object} options Options.
   *
   * Options list :
   * - headings: an array of headings to retrieve. Default : `['h2', 'h3', 'h4', 'h5', 'h6']`
   * - title: the title to print before the table of contents. Default: `Table of contents`.
   * - titleTag: the title tag. Default: `h2`.
   * - listType: the list type (`ul` or `ol`). Default: `ol`.
   */
  init(source, target, options) {
    this.setSource(source);
    this.setTarget(target);
    this.setOptions(options);
    this.printToc();
  }
}

const Minimalist_TOC = new TableOfContent();

/*
 * Here an example to init the table of contents without title:
 * - `content` is the id of the page content
 * - `toc` is the id where to place the table of contents.
 *
 * Minimalist_TOC.init('content', 'toc', {
	title: '',
});
 */
